### PR TITLE
CSB-48: Add weights to datasets

### PR DESCRIPTION
**Pull request by jan-Kulpas**

### CODE REVIEW

File: model_utilities/config.ini [hunk 1]
```diff
@@ -23,8 +23,14 @@ weight_decay = 1e-3
 scale_parameter = False
 relative_step = False
 
+[DATASET WEIGHTS]
```
Comment (filip-szweda) [inline model_utilities/config.ini]:
```suggestion
[DATA WEIGHTS]
```
Or change `[DATA]` in the config to `[DATASETS]`.

File: model_utilities/config.ini [hunk 2]
```diff
@@ -23,8 +23,14 @@ weight_decay = 1e-3
...
+dictionaries = 0.166
+corpus = 0.334
+remus = 0.5
+
+
 [DATA]
-training_data_file = ${DIRECTORIES:preprocessed_data_dir}/train.tsv
-validation_data_file = ${DIRECTORIES:preprocessed_data_dir}/val.tsv
+;training_data_file = ${DIRECTORIES:preprocessed_data_dir}/train.tsv
```
Comment (filip-szweda) [inline model_utilities/config.ini]: What are the `;` for?

File: model_utilities/train/data_loader.py [hunk 3]
```diff
@@ -1,3 +1,4 @@
+import os
```
Comment (filip-szweda) [inline model_utilities/train/data_loader.py:1]: `os` is a Python default library. Please separate it from external libraries with an empty line for better readability.

Comment (filip-szweda) [reply model_utilities/train/data_loader.py:1]: Please remember to run the pre-commit hook, if you haven't done so. You can run it with the `--all-files` flag.

File: model_utilities/train/data_loader.py [hunk 4]
```diff
@@ -19,3 +19,33 @@ def load_dataset(train_path: str, validation_path: str, test_path: str) -> ds.Da
     dataset = dataset.remove_columns("Unnamed: 0")
 
     return dataset
+
+
+def prepare_train_dataset(config: dict) -> ds.DatasetDict:
```
Comment (filip-szweda) [inline model_utilities/train/data_loader.py]: Pass only `config["DIRECTORIES"]["preprocessed_data_dir"]`, `config["TRAINING"]["shuffle_seed"]`, and `config.items("DATASET WEIGHTS")` as arguments. Load `config.items("DATASET WEIGHTS")` into a designated data structure beforehand to avoid operating on the entire config, as it violates the Single Responsibility Principle (SRP).

File: model_utilities/train/data_loader.py [hunk 5]
```diff
@@ -19,3 +19,33 @@ def load_dataset(train_path: str, validation_path: str, test_path: str) -> ds.Da
...
+    weights = {item[0]: float(item[1]) for item in config.items("DATASET WEIGHTS")}
+
+    paths = {key: {word: os.path.join(config["DIRECTORIES"]["preprocessed_data_dir"], f"{word}-{key}.tsv") for word in ["train", "val"]} for key in weights.keys()}
```
Comment (filip-szweda) [inline model_utilities/train/data_loader.py]: Move to a helper function e.g. `get_dataset_paths`.

File: model_utilities/train/data_loader.py [hunk 6]
```diff
@@ -19,3 +19,33 @@ def load_dataset(train_path: str, validation_path: str, test_path: str) -> ds.Da
...
+
+    # Need to interleave each split separately
+    train_datasets = {key: datasets[key]["train"] for key in datasets}
+
+    train_dataset_combined = ds.interleave_datasets(list(train_datasets.values()),
+                                                    list(weights.values()),
+                                                    seed=int(config["TRAINING"]["shuffle_seed"]),
+                                                    stopping_strategy="all_exhausted",
+                                                    )
+    val_datasets = {key: datasets[key]["validation"] for key in datasets}
```
Comment (filip-szweda) [inline model_utilities/train/data_loader.py]:
```suggestion
    validation_datasets = {key: datasets[key]["validation"] for key in datasets}
```
The term `val` can be misleading as it implies `value`, which it is not.

File: model_utilities/train/data_loader.py [hunk 7]
```diff
@@ -19,3 +19,33 @@ def load_dataset(train_path: str, validation_path: str, test_path: str) -> ds.Da
...
+
+
+def prepare_train_dataset(config: dict) -> ds.DatasetDict:
+    weights = {item[0]: float(item[1]) for item in config.items("DATASET WEIGHTS")}
+
+    paths = {key: {word: os.path.join(config["DIRECTORIES"]["preprocessed_data_dir"], f"{word}-{key}.tsv") for word in ["train", "val"]} for key in weights.keys()}
+
+    datasets = {key: load_dataset(paths[key]["train"], paths[key]["val"]) for key in paths}
+
+    # Need to interleave each split separately
```
Comment (filip-szweda) [inline model_utilities/train/data_loader.py]: This comment isn't helpful. The key question is "Why do we need to interleave each split separately?"

File: model_utilities/train/data_loader.py [hunk 8]
```diff
@@ -19,3 +19,33 @@ def load_dataset(train_path: str, validation_path: str, test_path: str) -> ds.Da
...
+    train_datasets = {key: datasets[key]["train"] for key in datasets}
+
+    train_dataset_combined = ds.interleave_datasets(list(train_datasets.values()),
+                                                    list(weights.values()),
+                                                    seed=int(config["TRAINING"]["shuffle_seed"]),
+                                                    stopping_strategy="all_exhausted",
+                                                    )
+    val_datasets = {key: datasets[key]["validation"] for key in datasets}
+
+    val_dataset_combined = ds.interleave_datasets(list(val_datasets.values()),
```
Comment (filip-szweda) [inline model_utilities/train/data_loader.py]:
```suggestion
    val_dataset_combined = ds.interleave_datasets(list(val_datasets.values()),
```
The term `val` can be misleading as it implies `value`, which it is not.

File: model_utilities/train/data_loader.py [hunk 9]
```diff
@@ -19,3 +19,33 @@ def load_dataset(train_path: str, validation_path: str, test_path: str) -> ds.Da
...
 
     return dataset
+
+
+def prepare_train_dataset(config: dict) -> ds.DatasetDict:
+    weights = {item[0]: float(item[1]) for item in config.items("DATASET WEIGHTS")}
+
+    paths = {key: {word: os.path.join(config["DIRECTORIES"]["preprocessed_data_dir"], f"{word}-{key}.tsv") for word in ["train", "val"]} for key in weights.keys()}
+
+    datasets = {key: load_dataset(paths[key]["train"], paths[key]["val"]) for key in paths}
```
Comment (filip-szweda) [inline model_utilities/train/data_loader.py:28]: I'm not a fan of this überwörterbuch approach. I think we should create separate vectors for each dataset set, such as `train_datasets`, `validation_datasets`, etc. and then merge them into `train_dataset`, `validation_dataset` etc.

File: model_utilities/train/data_loader.py (hunk 5 above)
Comment (filip-szweda) [reply model_utilities/train/data_loader.py]:
```suggestion
    paths = {key: {word: os.path.join(config["DIRECTORIES"]["preprocessed_data_dir"], f"{word}-{key}.tsv") for word in ["train", "validation"]} for key in weights.keys()}
```
The term `val` can be misleading as it implies value, which it is not.

File: model_utilities/config.ini (hunk 2 above)
Comment (jan-Kulpas) [reply model_utilities/config.ini]: commented out as it is unused in the new implementation

File: model_utilities/train/data_loader.py (hunk 3 above)
Comment (jan-Kulpas) [reply model_utilities/train/data_loader.py:1]: I did run it, weird that it wasn't caught.

File: model_utilities/train/data_loader.py (hunk 5 above)
Comment (jan-Kulpas) [reply model_utilities/train/data_loader.py]: We are generating the filenames of data to load here to avoid passing 9 different filepaths statically. We can't change this as that's how the files are named in pl-csb-data.

File: model_utilities/train/data_loader.py (hunk 7 above)
Comment (jan-Kulpas) [reply model_utilities/train/data_loader.py]: I'll edit this, it's because the Datasets library doesn't allow interleaving datasets containing splits

File: model_utilities/train/data_loader.py (hunk 6 above)
Comment (jan-Kulpas) [reply model_utilities/train/data_loader.py]: Okay, I agree about this here.

File: model_utilities/train/data_loader.py (hunk 9 above)
Comment (jan-Kulpas) [reply model_utilities/train/data_loader.py:28]: you might be right honestly, it made more sense when we had also separate test dataset for each source but that was changed.

File: model_utilities/config.ini (hunk 2 above)
Comment (filip-szweda) [reply model_utilities/config.ini]: Please remove the lines then. We do not need commented-out configuration, as it is already stored in the git history.

File: model_utilities/train/data_loader.py (hunk 9 above)
Comment (jan-Kulpas) [reply model_utilities/train/data_loader.py:28]: Changed my mind, our current implementation requires us to load a dataset with two splits by specifying both a train path and validation path and I don't want to untangle that as it provides little benefit anyway.

**Review by filip-szweda — APPROVED** [review]
